pkg/operator/ceph/nvmeof/controller.go

```diff
@@ -0,0 +1,624 @@
+			return errors.Wrapf(err, "failed to create/update deployment for %q", daemonID)
+		}
+
+		err = r.createCephNVMeOFService(cephNVMeOFGateway, daemonID)
+		if err != nil {
+			logger.Errorf("failed to create service for daemonID=%s: %v", daemonID, err)
+			return errors.Wrapf(err, "failed to create service for %q", daemonID)
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcileCephNVMeOFGateway) downCephNVMeOFGateway(cephNVMeOFGateway *cephv1.CephNVMeOFGateway, currentCount int) error {
+	for i := cephNVMeOFGateway.Spec.Instances; i < currentCount; i++ {
+		daemonID := fmt.Sprintf("%d", i)
+		name := instanceName(cephNVMeOFGateway, daemonID)
+
+		err := r.context.Clientset.AppsV1().Deployments(cephNVMeOFGateway.Namespace).Delete(r.opManagerContext, name, metav1.DeleteOptions{})
+		if err != nil && !kerrors.IsNotFound(err) {
+			logger.Errorf("failed to delete deployment %s: %v", name, err)
+			return errors.Wrapf(err, "failed to delete deployment %q", name)
+		}
+
+		err = r.context.Clientset.CoreV1().Services(cephNVMeOFGateway.Namespace).Delete(r.opManagerContext, name, metav1.DeleteOptions{})
+		if err != nil && !kerrors.IsNotFound(err) {
+			logger.Errorf("failed to delete service %s: %v", name, err)
+			return errors.Wrapf(err, "failed to delete service %q", name)
+		}
+	}
+
+	return nil
+}
+
+// getNVMeOFGatewayConfig generates a complete nvmeof.conf configuration file
+// with all values filled in (no placeholders). User overrides from nvmeofConfig
+// are merged on top of the default configuration.
+func getNVMeOFGatewayConfig(poolName, podName, podIP, anaGroup string, userConfig map[string]map[string]string) (string, error) {
+	cfg := ini.Empty()
+	// Set default [gateway] section
+	gatewaySection, err := cfg.NewSection("gateway")
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create gateway section")
+	}
+	gatewaySection.Key("name").SetValue(podName)
+	gatewaySection.Key("group").SetValue(anaGroup)
+	gatewaySection.Key("addr").SetValue(podIP)
+	gatewaySection.Key("port").SetValue("5500")
+	gatewaySection.Key("enable_auth").SetValue("False")
+	gatewaySection.Key("state_update_notify").SetValue("True")
+	gatewaySection.Key("state_update_timeout_in_msec").SetValue("2000")
+	gatewaySection.Key("state_update_interval_sec").SetValue("5")
+	gatewaySection.Key("enable_spdk_discovery_controller").SetValue("False")
+	gatewaySection.Key("encryption_key").SetValue("/etc/ceph/encryption.key")
+	gatewaySection.Key("rebalance_period_sec").SetValue("7")
+	gatewaySection.Key("max_gws_in_grp").SetValue("16")
+	gatewaySection.Key("max_ns_to_change_lb_grp").SetValue("8")
+	gatewaySection.Key("verify_listener_ip").SetValue("False")
+	gatewaySection.Key("enable_monitor_client").SetValue("True")
+
+	// Set default [gateway-logs] section
+	gatewayLogsSection, err := cfg.NewSection("gateway-logs")
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create gateway-logs section")
+	}
+	gatewayLogsSection.Key("log_level").SetValue("debug")
```

Comment:
Should we consider changing the default based on ROOK_LOG_LEVEL? Thoughts @travisn ?